### ISSUE TITLE
Hayden Phet Statistics Independent Study

### ISSUE THREAD

**Issue by Berge17** [phetsims/website#221]
# Initial findings

After becoming familiar with the basic structure of the server and reviewing the readme of the third party tool, here is what I have concluded so far in layman's terms:

1) Logs for each month must be combined into a single file before the script is run to create the sql commands

2) Logs (that are used in the readme) are located in data/web/htdocs/phetsims/statistics/month_sort2

3) Each directory under month_sort2 contains different types of logs (combined, current, etc...)

- I believe I have gotten a good idea of the general process that the tool uses.

# Questions to investigate/Next Steps

1) Where are the rest of the logs located that haven't been input yet? (I know where they are but want to document this more)

2) Familiarize myself with the sql DB 

- Once I complete these tasks I believe that I can get a better understanding of how the tool works, use it to input new data, and then document the tool further. I will update the tasks as I progress

**Comment by Berge17** [comment]
# Update 6/3

# Steps to obtain data

In order for the tool to be used to create statements for the data to be entered into the DB, it must first be in the proper form. Here is a summary of the steps taken to prepare the data.

- data2/figaro-logs/apache has the logs in zipped form.

1) Copy the relevant files (files for that month) to another directory /data2/spot-check-analysis is an example of this

2) Unzip the files

3) Cat the relevant files to a single file (i.e. access logs from May to an access_May file). Grep file for relevant portion and Grep -v to omit 404's. This is a very crucial step along with step 4.

4) Specific simulations can be extracted out using cat and egrep

# Conclusion

I made good progress on the process of preparing the data, tomorrow I will examine the current sql DB. Some of the combined logs made in the Readme are not simulation specific; they contain all logs of simulations for that month. While the directions in the issue only use a specific simulation. I will need to look in the DB and the perl script to see how the DB is laid out if cating and greping for a specific sim is necessary of if they all can be entered at once.

**Comment by Berge17** [comment]
# Update

Connected to sql database using following command: /usr/local/mysql/bin/mysql -u phet_sims phet_sims -p 

password is found in config.php

below is a picture of the tables 

![screen shot 2015-06-08 at 2 54 42 pm](https://cloud.githubusercontent.com/assets/5994594/8045112/532abac6-0dee-11e5-9ede-e9acd3a4ad40.png)

I have looked around at these and the next step will be to learn about how the perl script works with each table and document this, then insert the new data into each table as necessary.

**Comment by Berge17** [comment]
# Update

In the perl script create_sims_insert.pl, which is the script used to create insert statements, the insert statements are inserted into the sim_counts table. I am not familiar with perl but I am familiarizing myself with it and can understand enough to know what the script is doing. 

This is an example of what the combined  log file should look like

![screen shot 2015-06-10 at 4 07 45 pm](https://cloud.githubusercontent.com/assets/5994594/8095398/1c95f9b4-0f8b-11e5-8ea5-f50ee7e2082b.png)

This is what the insert statement should look like after the perl script is ran:

INSERT INTO sim_counts(year,month,count,project,flavor,language,codetype,simtype,logtype,version,html5)
VALUES (2013,4,14225,'acid-base-solutions','acid-base-solutions','all','jar','sims','sims','',0)

The next step for me is to create the combined files for the new data and use the perl script to make the insert statements, then compare the insert statements to the data in order to make sure they are correct. After that I can insert the data in the db.

**Comment by Berge17** [comment]
# Update

Trying to get php to work on command line. Have to edit -./bashrc as Brian stated in other issue. Will document this once I get it working.

**Comment by ariel-phet** [comment]
@Berge17 - once you feel you have a pretty good understanding of the system (and you might already be at that point) - it might be productive to have you meet with one of our developers that was the liaison to the 3rd party which created this code ( @jonathanolson ). Let me know and we will set up a meeting with him.

**Comment by Berge17** [comment]
@ariel-phet A meeting sounds like a great idea. I am familiar with the system and feel ready for the next step. I can meet almost any time during the week.

**Comment by ariel-phet** [comment]
@Berge17 - can you summarize your meeting with @jonathanolson here and comment on next steps? Thanks!

**Comment by Berge17** [comment]
# Summary of Meeting 6/16

- Overview/High level documentation a good idea (Directories).
- Stats are pulled from 3 main sources: apache logs (java), offline use database, and google analytics (html5).
- Most important is to insert new statistics. Specifically access logs and downloads from apache logs. Error logs not so much.
- Use what is written in issue 235 tasks to check and see if data is accurate.
- Watch out for redundancy (same data multiple times).
- Jonathan will be available to answer more questions.

# Next steps

Look at php files. Document log directories/directory structure. Try to insert data. Check data using steps in issue to make sure it is accurate.

**Comment by ariel-phet** [comment]
Thanks @Berge17 - for the next steps can you put in some estimates/milestones? It would be good to know where you think you can be by the end of the week.

**Comment by Berge17** [comment]
# Update

Important Files

# WeeklyLoad.php
- Function loadWeekly uses Loader class (Loader.php). Creates new loader with ('~/tmp-logs') as an argument.
- uses function loadall from Loader.php

# Loader.php
- This file contains the loader class and lots of functions. Most importantly the writeall and loadall functions.
-  $logfile variable comes from config.php file
- parser functions are also used. Located in parse.php

# Parse.php
- Loads logs into proper format by correctly parsing them. 

# Goals for the week:
1- high level documentation of system/directories (mostly log files/locations)
2- Further Documentation on 3 above files
3- Possibly write file to test Loader
4- Begin making use of tool and preparing logs to be input

**Comment by Berge17** [comment]
# Update

# Loader.php
- loadall function.
-- for each file in directory (directory is specified in config.php) file is loaded and parsed. This does not create an sql statement of load anything into db. Simply parses file and returns an array of all parsed lines ( will test later tonight).

- writeLines is a function that actually writes the sql statements. I believe writeLines can be executed once a file has been parsed. Writelines checks date of sim to make sure it has not been loaded yet.

I need to find what file the statements are written to. (Path may have to be changed)

**Comment by Berge17** [comment]
#WriteLines

Works and parses file correctly. Returns last element of the parsed Array. I believe this is correctly entering data into database. I will test this tomorrow and write a few output statements. I have changed LoadWeekly.php to now use writeLines instead of Loadall.

**Comment by Berge17** [comment]
# Update

Over this weekend I made some significant progress. I found that writelines writes the correct sql lines to a file (sims.sql) in the tmp directory. However this data is not automatically loaded into the sql sims table. I will double check but as of now the sql statements seem to be correct. The next step is to find a way to update the table's contents with the new data. 

There is a function called insertFromFile that can insert sql from a file to a table. Once the data is checked I believe it can be loaded using this function.

**Comment by ariel-phet** [comment]
@Berge17 - thanks for the weekend update!

Can you put some milestones for the week, and also remind me of the end date for your independent study? I think it would be good to get the new data loaded into the data base this week and it sounds like you are almost there.

**Comment by Berge17** [comment]
# Goals for week of 6/22

1) Write a script to import sql file into database
2) Test script on larger log file
3) Double check writeLines function is writing proper statements

After I finish these I will import all the new data and write a comprehensive report about how the system works, what I learned about it, and where the process can likely be improved. 

The end date for the independent study is July 2nd.

**Comment by Berge17** [comment]
# importHelper.php

For some reason, the given function the tool uses to import an sql file into the database does not work. I have added a php file importHelper.php that imports the file tmp/sims.sql into the database. sims.sql is the file that the tool writes sql statements to.

**Comment by ariel-phet** [comment]
@Berge17 - please make an update at the end of today. Glad to see almost all of this week's goals are almost complete!

**Comment by Berge17** [comment]
# Update 6/26

As of right now, everything is ready to import data into the database. The last step is to make sure the data that is being entered into the system is the correct data. I am meeting with Jonathan on Tuesday to make sure it is all correct and after that I will update the database. 

This is the process right now.

1) Input file to be loaded in WeeklyLoad.php, which writes corresponding input statements to sims.sql located in the tmp file.
2) Run importHelper.php to import sql statements into database.
3) Clear sims.sql to make sure data is not inserted twice. 

- This process could easily be changed to one step and also to insert full files from directories instead of a single file at a time. However, running WeeklyLoad for one log file takes about 20 minutes and I think right now the steps are a good way of double checking everything is running correctly. As far as redundancy goes there is an easy sql query to find duplicate rows which I will document further this weekend. 

# Goals for this weekend:

1) Continue to test that sql statements are accurate
2) Find what all files/data is already in the database and document it
3) Devise a solution to check for data redundancy

**Comment by Berge17** [comment]
# Weekend Update

1) I do believe the sql statements are correct to the best of my knowledge. 

2) Most recent logs in database are from 8/14. Earliest ID's are from 06/23. I haven't found any holes yet but I may.  

3) I believe this can be done using the COUNT function in sql, however when I tried to use it the query also took a very long time.

**Comment by Berge17** [comment]
# Update 7/1

I have begun entering data. Right now I am doing the access logs and if I have time I will also do ssl and downloads. I am using the php tool to insert into the sims table. I ran into a few issue due to memory being full but have worked around it for now. I have commented this in the code in WeeklyLoad.php and Loader.php. I will update again later this afternoon

**Comment by Berge17** [comment]
# Update 2 7/1

Trying to write statements in bulk did not work. After 3 hours I was kicked off of ssh broken (pipe error) apparently there isn't much to do about that.  Tried to write the statements that were written during that time to database and got a sql is gone error. This can be remedied by changing some sql settings but I do not have permission to do so. I will try doing each log individually tomorrow, that was the way I had success before. This will also take a long time but should work. If I can't get this to work I will prepare the logs as best I can and include the steps for someone with sudo permission to change the settings for a bulk import. The sheer size of these logs makes it difficult to do in bulk. As I said before none of the for statements or other bulk strategies I used work due to the size of the logs.

**Comment by Berge17** [comment]
# Update 7/2

This is the last day of my independent study, I have tried numerous ways to deal with the large size of these logs and have been unsuccessful in each. I did find a way to insert each file individually however and this is documented in my final report. Basically, each file takes about 2 hours insert but I tried a few different ways of bulk inserting and these did not work (kicked of off ssh or mysql timed out). Below is the instructions to insert each log one by one.

	1) Be sure the sql statements located in /tmp/sims.sql have been cleared to avoid any duplicate inserts.
	2) Change the file path on line   of WeeklyLoad.php to the file you want to insert. Unzipped access logs are in data2/logdecomp
	3) Run WeeklyLoad.php using the command php WeeklyLoad.php
	4) (Optional) check /tmp/sims.sql to make sure statements have been written
	5) Run importHelper.php using the command php importHelper.php  

This is for inserts into the sims table. For sim_count there is another way whose instructions are located at /data/web/htdocs/phetsims/statistics/month_sort2/combine/README see issue 235 on tasks for more instructions.

The main code I worked on is found in WeeklyLoad.php and importHelper.php. If anyone has any questions about my work please feel free to contact me. I will be emailing Ariel a final report later today.

**Comment by jonathanolson** [comment]
Thanks! Closing.

**Comment by zepumph** [comment]
@ariel-phet it might be good to put a copy of the final report on this issue, if that's appropriate, just to have everything in one place.
